### ISSUE TITLE
Setting attributes of class instance in `__new__` method included in `__slots__` causes those attributes to be inferred as `Unbound`

### ISSUE THREAD

**Issue by jonathandung** [microsoft/pyright#11373]
**Describe the bug**
When a class defines a `__new__` method rather than `__init__` and uses `__slots__` (e.g. for immutable objects with a fixed structure), the attributes set on the return object, even if known to be an instance of that class, in its body, are not recognized to have been 'bound' if not re-set elsewhere, and are revealed to be `Unbound` when accessed in other methods of the class.

**Code or Screenshots**

```python
from _thread import LockType
from typing import reveal_type
class Foo:
    __slots__ = '_spam', '_ham'
    def __new__(cls, tup: tuple[int, LockType]):
        self = super().__new__(cls)
        self._spam, self._ham = tup
        return self
    def bar(self): reveal_type(self._spam) # no diagnostic produced
    # when I hover over the attribute, it says 'Unbound'
    def baz(self):
        with self._ham: # ERROR!
             self.bar()
```

The error shown in the above case is twofold:

```md
Object of type "Unbound" cannot be used with "with" because it does not correctly implement __enter__ Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/docs/diagnostics/reportGeneralTypeIssues.md)
Object of type "Unbound" cannot be used with "with" because it does not correctly implement __exit__ Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/docs/diagnostics/reportGeneralTypeIssues.md)
```

**VS Code extension or command-line**
VS Code extension; preview (v2026.2.100).
I have seen analogs of this issue in previous versions as well.
